src/net/socket.rs

```diff
@@ -1,87 +1,94 @@
-            if !link.should_send() {
-                return Ok(0);
+}
+/// Provides a `SocketSender` implementation for `SocketWithConditioner`
+impl SocketSender for SocketWithConditioner {
+    // When `LinkConditioner` is enabled, it will determine whether packet will be sent or not.
+    fn send_packet(&mut self, addr: &SocketAddr, payload: &[u8]) -> Result<usize> {
+        if cfg!(feature = "tester") {
+            if let Some(ref mut link) = *self
+                .link_conditioner
+                .lock()
+                .expect("Lock is already held by current thread.")
+            {
+                if !link.should_send() {
+                    return Ok(0);
+                }
             }
         }
         Ok(self.socket.send_to(payload, addr)?)
     }
+}
 
-    /// Returns mutable reference of `UdpSocket`
-    pub fn socket(&mut self) -> &mut UdpSocket {
-        &mut self.socket
-    }
-
-    /// Returns the local socket address
-    pub fn local_addr(&self) -> Result<SocketAddr> {
-        Ok(self.socket.local_addr()?)
+/// Provides a `SocketReceiver` implementation for `UdpSocket`
+impl SocketReceiver for UdpSocket {
+    /// Receive a single packet from UDP socket.
+    fn receive_packet<'a>(
+        &mut self,
+        buffer: &'a mut [u8],
+    ) -> Result<Option<(&'a [u8], SocketAddr)>> {
+        Ok(match self.recv_from(buffer) {
+            Ok((recv_len, address)) => {
+                if recv_len == 0 {
+                    return Err(ErrorKind::ReceivedDataToShort);
+                }
+                Some((&buffer[..recv_len], address))
+            }
+            Err(e) => {
+                if e.kind() != std::io::ErrorKind::WouldBlock {
+                    return Err(e.into());
+                }
+                None
+            }
+        })
     }
-
-    /// Set the link conditioner for this socket. See [LinkConditioner] for further details.
-    pub fn set_link_conditioner(&mut self, conditioner: Option<LinkConditioner>) {
-        self.link_conditioner = conditioner;
+    /// Returns the socket address that this socket was created from.
+    fn local_addr(&self) -> Result<SocketAddr> {
+        Ok(self.local_addr()?)
     }
 }
 
 /// A reliable UDP socket implementation with configurable reliability and ordering guarantees.
 #[derive(Debug)]
 pub struct Socket {
-    socket_wrapper: SocketWithConditioner,
-    config: Config,
-    connections: ActiveConnections,
-    recv_buffer: Vec<u8>,
-    event_sender: Sender<SocketEvent>,
-    packet_receiver: Receiver<Packet>,
-
-    receiver: Receiver<SocketEvent>,
-    sender: Sender<Packet>,
-}
-
-enum UdpSocketState {
-    MaybeEmpty,
-    MaybeMore,
+    // Stores an instance of `SocketController` where `SocketSender` uses a `UdpSocket` (with `LinkConditioner`, if enabled) and SocketReceiver` is a `UdpSocket`.
+    handler: SocketController<SocketWithConditioner, UdpSocket>,
+    link_conditioner: Arc<Mutex<Option<LinkConditioner>>>,
```

Comment:
I tried to delegate it, but it looks "incorrect" 
I need to pass delegate it through a SocketImpl as well as ConnectionController, and they do not care at all about this link_conditioner. I see that we have two (three?) options here:
1) delegate through all interfaces, regardless if they care about it or not.
PRO: no performance penalty when a feature is enabled.
CON: the need to change interfaces, that doesn't care about this.
2) use `Arc<Mutex<....>>`
PRO: cleaner interfaces
CON: performance penalty when a feature is enabled.
3) Since we know that `LinkConditioner` will never be set from a different thread, maybe we can express that in rust, but I need your help here, I'm not very proficient in rust yet...
PRO: no (or minimum) performance penalty when a feature is enabled.
CON: ... maybe it might require to write "unsafe" code.

What should I do?